src/pre_process/m_initial_condition.fpp

```diff
@@ -283,6 +285,1204 @@ contains
+				ii = 3; jj = 5; ci((ii-1)*(n+1)+j,(jj-1)*(n+1)+k) = -d(j,k)/(gam*mach**2);
+				ii = 5; jj = 3; ci((ii-1)*(n+1)+j,(jj-1)*(n+1)+k) = -(gam-1)*d(j,k)/rho_mean(j);
+			end do
+		end do
+		ar = br
+		ai = bi+ci
+
+		! Compute Eigenvalues & Eigenfunctions
+		call cg(5*(n+1),5*(n+1),ar,ai,wr,wi,zr,zi,fv1,fv2,fv3,ierr)
+
+		! Find the most unstable wave
+		call find_unstable_mode(5*(n+1),wr,wi,zr,zi,tr,ti,vr,vi)
+
+		! Normalize
+		call normalize_eigvec(5*(n+1),vr,vi,vnr,vni)
+
+		! Generate instability waves
+		call generate_wave(5*(n+1),vnr,vni,alpha,beta,wave,shift)
+
+    end subroutine s_instability_wave
+
+	subroutine generate_wave(nl,vnr,vni,alpha,beta,wave,shift)
+		integer nl
+		real(kind(0d0)) :: alpha,beta,ang
+		real(kind(0d0)), dimension(0:nl-1) :: vnr,vni
+		real(kind(0d0)), dimension(5,0:m,0:n,0:p) :: wave
+		real(kind(0d0)) :: shift
+		integer i,j,k
+		
+		do i=0,m
+		do j=0,n
+		do k=0,p
+			if (beta .eq. 0) then
+				ang = alpha*x_cc(i)
+			else 
+				ang = alpha*x_cc(i)+beta*z_cc(k)+shift
+			end if
+			
+			wave(1,i,j,k) = vnr(j)*cos(ang) &
+						   -vni(j)*sin(ang)		! rho
+			wave(2,i,j,k) = vnr((n+1)+j)*cos(ang) &
+						   -vni((n+1)+j)*sin(ang)	! u
+			wave(3,i,j,k) = vnr(2*(n+1)+j)*cos(ang) &
+						   -vni(2*(n+1)+j)*sin(ang)	! v
+			wave(4,i,j,k) = vnr(3*(n+1)+j)*cos(ang) &
+						   -vni(3*(n+1)+j)*sin(ang)	! w
+			wave(5,i,j,k) = vnr(4*(n+1)+j)*cos(ang) &
+						   -vni(4*(n+1)+j)*sin(ang)	! T
+		end do
+		end do
+		end do
+	
+	end subroutine generate_wave
+	
+	subroutine normalize_eigvec(nl,vr,vi,vnr,vni)
+	integer nl
+	real(kind(0d0)), dimension(nl) :: vr,vi,vnr,vni
+	real(kind(0d0)) :: norm
+	real(kind(0d0)) :: tr,ti,cr,ci
+	integer i,idx
+	
+	norm = 0d0
+	do i=1,nl
+		if (dsqrt(vr(i)**2+vi(i)**2) .gt. dabs(norm)) then
+			idx = i
+			if (vr(i) .gt. 0) then 
+				norm = sqrt(vr(i)**2+vi(i)**2)
+			else
+				norm =-sqrt(vr(i)**2+vi(i)**2)
+			end if
+		end if
+	end do
+
+	vnr = vr/norm
+	vni = vi/norm
+
+	tr = vnr(idx)
+	ti = vni(idx)
+	do i=1,nl
+		call cdiv(vnr(i),vni(i),tr,ti,cr,ci)
+		vnr(i) = cr
+		vni(i) = ci
+	end do
+	
+	end subroutine normalize_eigvec
+
+	subroutine find_unstable_mode(nl,wr,wi,zr,zi,tr,ti,vr,vi)
+		integer nl
+		real(kind(0d0)), dimension(nl) :: wr,wi
+		real(kind(0d0)), dimension(nl,nl) :: zr,zi
+		real(kind(0d0)) :: tr,ti
+		real(kind(0d0)), dimension(nl) :: vr,vi
+		integer i,k
+		
+		k=1
+		do i=2,nl
+			if (wi(i) .gt. wi(k)) then
+				k = i
+			end if
+		end do
+	
+		tr = wr(k)
+		ti = wi(k)
+		vr = zr(:,k)
+		vi = zi(:,k)
+	
+	end subroutine find_unstable_mode
+
+!====================================================================
+!	Subroutines & functions for computing eigenvalues and eigenvectors
+!	which are modified from EISPACK (https://netlib.org/eispack/)
+!		1) cg
+!		2) cbal
+!		3) corth
+!		4) comqr2
+!		5) csroot
+!		6) cdiv
+!		7) pythag
+!====================================================================
+	subroutine cg(nm,nl,ar,ai,wr,wi,zr,zi,fv1,fv2,fv3,ierr)
```

Comment:
these routines should live in a separate module so they do not pollute this one. probably in the `commmon/` directory.

Reply:
All subroutines for solving the eigenvalue problem are moved to m_eigen_solver.f90 in common directory.